internal/controller/controllers/agent_controller.go

```diff
@@ -399,15 +452,25 @@ func (r *AgentReconciler) handleAgentFinalizer(ctx context.Context, log logrus.F
 		}
 	} else { // agent is being deleted
 		if funk.ContainsString(agent.GetFinalizers(), AgentFinalizerName) {
-			if _, has_annotation := agent.GetAnnotations()[AgentSpokeCleanupAnnotation]; has_annotation && agent.Spec.ClusterDeploymentName != nil {
+			if _, skipCleanup := agent.GetAnnotations()[AgentSkipSpokeCleanupAnnotation]; !skipCleanup && agent.Spec.ClusterDeploymentName != nil {
 				// only remove the spoke resources if the entire cluster isn't being deleted
-				clusterExists, err := r.clusterExists(ctx, agent)
+				clusterRef := types.NamespacedName{
+					Name:      agent.Spec.ClusterDeploymentName.Name,
+					Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+				}
+				clusterExists, err := r.clusterExists(ctx, clusterRef)
 				if err != nil {
 					log.WithError(err).Errorf("failed to check cluster deployment presence")
 					return &ctrl.Result{}, err
 				}
 				if clusterExists {
-					if err := r.removeSpokeResources(ctx, log, agent); err != nil {
+					spokeClient, err := r.spokeKubeClient(ctx, agent.Spec.ClusterDeploymentName)
+					if err != nil {
+						log.WithError(err).Error("failed to create spoke client")
+						return &ctrl.Result{}, err
+					}
+					nodeName := getAgentHostname(agent)
+					if err := removeSpokeResources(ctx, log, spokeClient, nodeName); err != nil {
```

Comment:
Not in the scope of your PR but maybe `removeSpokeResources` could be revisited?

Could a [machine not existing error here](https://github.com/openshift/assisted-service/blob/f31e7f2f0ba0b00c696f5cfc7ff2c3f076b45e6b/internal/controller/controllers/agent_controller.go#L387-L389) cause this Agent resource to hang here?